### PR TITLE
Move definition of column colors etc. into a new type

### PR DESCRIPTION
**Pull request by mcendu**
Prereq for skinning mania.

### CODE REVIEW

File: osu.Game.Rulesets.Mania/UI/Components/EvenColumn.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osuTK.Graphics;
+
+namespace osu.Game.Rulesets.Mania.UI.Components
+{
+    public class EvenColumn : ColumnType
```
Comment (mcendu) [inline osu.Game.Rulesets.Mania/UI/Components/EvenColumn.cs]: `ColumnTypeEven` might be better.

File: osu.Game.Rulesets.Mania/UI/Components/ColumnType.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public abstract Color4 ColumnColour { get; }
+        public virtual float DefaultColumnWidth => Column.COLUMN_WIDTH;
+
+        public virtual string NoteTexture => "mania-note1";
+        public virtual string HoldTexture => "mania-note1L";
+        public virtual string HoldHeadTexture => "mania-note1H";
+        public virtual string HoldTailTexture => "mania-note1T";
+        public virtual string KeyTexture  => "mania-key1";
+        public virtual string KeyDownTexture => "mania-key1D";
```
Comment (mcendu) [inline osu.Game.Rulesets.Mania/UI/Components/ColumnType.cs]: These values does not point to the actual textures but the name of the textures.

Another name could be better.

File: osu.Game.Rulesets.Mania/UI/Components/EvenColumn.cs (hunk 1 above)
Comment (iiSaLMaN) [reply osu.Game.Rulesets.Mania/UI/Components/EvenColumn.cs]: As far as I've seen, derived class name mostly comes before inherited class name, sometimes in-between and can be shortened if not misleading:
    - `FooType : Type`
    - `DrawableFooType : DrawableType`

File: osu.Game.Rulesets.Mania/UI/Components/ColumnType.cs (hunk 2 above)
Comment (iiSaLMaN) [reply osu.Game.Rulesets.Mania/UI/Components/ColumnType.cs]: perhaps `...TextureName`?

File: osu.Game.Rulesets.Mania/UI/Components/OddColumnType.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osuTK.Graphics;
+
+namespace osu.Game.Rulesets.Mania.UI.Components
+{
+    public class OddColumnType : ColumnType
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Mania/UI/Components/OddColumnType.cs]: I don't like how this column type is inconsistent conpared to other types, how about abstracting the texture name properties and override here?

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Mania/UI/Column.cs [hunk 4]
```diff
@@ -101,22 +100,25 @@ public Column(int index)
...
-                isSpecial = value;
+                columnType = value;
 
-                Width = isSpecial ? special_column_width : COLUMN_WIDTH;
+                AccentColour = columnType.ColumnColour;
+                Width = columnType.DefaultColumnWidth;
             }
         }
 
+        public bool IsSpecial { get; set; }
```
Comment (bdach) [inline osu.Game.Rulesets.Mania/UI/Column.cs]: Do we even want this to still be a thing? Maybe it shouldn't exist anymore, or at least be a get-only convenience property:
```csharp
public bool IsSpecial => columnType is SpecialColumnType;
```
I'm worried if this survives it'll be bug-prone (two sources of truth for one piece of info).

File: osu.Game.Rulesets.Mania/UI/Components/SpecialColumnType.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+// See the LICENCE file in the repository root for full licence text.
+
+using osuTK.Graphics;
+
+namespace osu.Game.Rulesets.Mania.UI.Components
+{
+    public class SpecialColumnType : ColumnType
+    {
+        public override Color4 ColumnColour => new Color4(0, 48, 63, 255);
+        public override float DefaultColumnWidth => 70;
```
Comment (bdach) [inline osu.Game.Rulesets.Mania/UI/Components/SpecialColumnType.cs]: I'd expose and reuse
https://github.com/ppy/osu/blob/81d2487f65f827a35b124719507d641e1da07285/osu.Game.Rulesets.Mania/UI/Column.cs#L25

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Mania/UI/Components/SpecialColumnType.cs (hunk 5 above)
Comment (mcendu) [reply osu.Game.Rulesets.Mania/UI/Components/SpecialColumnType.cs]: Is there any reason to still put it in `Column`?

Comment (bdach) [reply osu.Game.Rulesets.Mania/UI/Components/SpecialColumnType.cs]: IMO the two constants should be side by side for easy searching and access.

File: osu.Game.Rulesets.Mania/UI/Components/ColumnType.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osuTK.Graphics;
+
+namespace osu.Game.Rulesets.Mania.UI.Components
+{
+    public abstract class ColumnType
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/UI/Components/ColumnType.cs]: `ColumnType` is a bad name for this. `ColumnInfo` maybe?

Also please add full xmldoc for the class and all properties.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
I'm not sure this is the way we want to go about this. When considering skinning, most of the values here pertain only to legacy values (which will all be consolidated within `ManiaLegacySkinTransformer` (see the other similar classes for reference)).

Rather than giving columns a type, how about deriving the column to eventually result in an inheritance structure like:
```
Column:
    - LegacyColumn:
        - LegacyEvenColumn
        - LegacyOddColumn
        - LegacySpecialColumn
    - DefaultColumn:
        - DefaultEvenColumn
        - DefaultOddColumn
        - DefaultSpecialColumn
```

Then, all the `mania-noteX` members will only appear within `LegacyColumn`.

**Comment by mcendu** [comment]
Superseded by #7431 (thanks @smoogipoo for the hint).

**Pull request closed without merging**